### PR TITLE
Optimize memory initialization handling in AOT loader

### PR DESCRIPTION
**Pull request by no1wudi**
Save memory if the file buffer is always exist before exit.

### CODE REVIEW

**Comment by wenyongh** [comment]
This is to make the memory init data available and wasm/aot module can be used again when `module->is_binary_freeable` is true and the wasm/aot binary is freed after loading?

**Comment by no1wudi** [comment]
Yes, the AOT binary should be available before runtime exit

File: core/iwasm/aot/aot_loader.c [hunk 1]
```diff
@@ -1026,8 +1030,21 @@ load_mem_init_data_list(const uint8 **p_buf, const uint8 *buf_end,
         data_list[i]->offset.init_expr_type = init_value.init_expr_type;
         data_list[i]->offset.u = init_value.u;
         data_list[i]->byte_count = byte_count;
-        read_byte_array(buf, buf_end, data_list[i]->bytes,
-                        data_list[i]->byte_count);
+        data_list[i]->bytes = NULL;
+        if (module->is_binary_freeable) {
```
Comment (wenyongh) [inline core/iwasm/aot/aot_loader.c:1034]: There is similar handling in wasm loader, but it clones memory init data only when both `is_load_from_file_buf` and `module->is_binary_freeable` are true:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/interpreter/wasm_loader.c#L5791
note `is_load_from_file_buf` is passed to this fuction's caller `load_init_data_section`.

Do you think here we had better handle `data_lists[i]->bytes` like wasm_loader.c?

Comment (no1wudi) [reply core/iwasm/aot/aot_loader.c:1034]: Yes, it's better to handle them by same way

Comment (no1wudi) [reply core/iwasm/aot/aot_loader.c:1034]: But, pass the `is_load_from_file_buf ` will change many places, I think `module->is_binary_freeable` is enough for many use case

**Review by wenyongh — APPROVED** [review]
LGTM

File: core/iwasm/compilation/aot.c [hunk 2]
```diff
@@ -60,27 +63,39 @@ aot_create_mem_init_data_list(const WASMModule *module)
...
         data_list[i]->is_passive = module->data_segments[i]->is_passive;
         data_list[i]->memory_index = module->data_segments[i]->memory_index;
 #endif
         data_list[i]->offset = module->data_segments[i]->base_offset;
         data_list[i]->byte_count = module->data_segments[i]->data_length;
-        memcpy(data_list[i]->bytes, module->data_segments[i]->data,
-               module->data_segments[i]->data_length);
+        data_list[i]->bytes = NULL;
+        /* Allocate memory for AOT compiler is OK, because the data segment
+         * is small and the host memory is enough */
```
Comment (lum1n0us) [inline core/iwasm/compilation/aot.c:82]: If that's the case, why not maintain the original? But still, the modification was made.

**Review by TianlongLiang — APPROVED** [review]
LGTM

File: core/iwasm/compilation/aot.c (hunk 2 above)
Comment (lum1n0us) [reply core/iwasm/compilation/aot.c:82]: Are you suggesting that malloc()/free() should better match the 
implied convention of the data structure? If using a variable-length array as we did previously, we should allocate one block of memory. If using a pointer as we do now, we should allocate memory twice.

Comment (no1wudi) [reply core/iwasm/compilation/aot.c:82]: Using variable-length arrays is not an issue; it's just that in the current scenario, most of the time we don't even need to allocate memory a second time. We can simply point the pointer to the data buffer.

Comment (lum1n0us) [reply core/iwasm/compilation/aot.c:82]: I completely agree the intent of this PR. 

My uncertainty lies with the changes made in aot.c. As you mentioned, the AOT compiler doesn't need to *set the pointer to the data buffer*. From my perspective, this suggests that there is no need to alter aot.c, as it does not align with the same intention as aot_loader.c. But if this is about upholding an implied convention of programming, then I believe it is acceptable.

Comment (no1wudi) [reply core/iwasm/compilation/aot.c:82]: The key point is the structure that `uint8* bytes; // NEW` at the end can not be used with variable-length, it's totally different with `uint8 bytes[1]`, so I must modifty the `aot.c`.

Comment (lum1n0us) [reply core/iwasm/compilation/aot.c:82]: Oh, I see.

**Review by lum1n0us — APPROVED** [review]
LGTM

**Pull request merged**